### PR TITLE
Themes: Upload page tweaks

### PR DESCRIPTION
**Pull request by seear**
**1) Fix react warnings on mount**

screen shot 2016-12-14 at 12 39 53

Test: 
* Go to http://calypso.localhost:3000/design/{site}
* Click 'upload theme' button
* should not see the specific warning above in the console

**2) Fix back button**
Test:
* as above
* back button should always work

**3) Automatically close notices**
Test:
* as above
* upload a theme (must be a jetpack site)
* Success/error notice should disappear after 5 seconds

### CODE REVIEW

**Comment by folletto** [comment]
1. No warning on the .com "Upload Theme" page — ✅ *
2. No warning on the Jetpack "Upload Theme" page — ✅ *
3. .com site → Upload page → Back — ✅ 
4. Jetpack site → Upload page → Back — ✅  
5. Jetpack site → Upload page → Upload... → Upload done → Back — ✅ 
6. .com site → Upload page → switch to Jetpack site → Back — ❌ 
7. Success/error notice should disappear after 5 seconds — ✅ *

*(1, 2) However, I don't see that warning in `wpcalypso` either.

*(6) Can we make only the success message to disappear, and let the error messages stay?

Q— 
What happens if I browse away (back) while it's uploading?

**Comment by seear** [comment]
> *(1, 2) However, I don't see that warning in wpcalypso either.

The warnings are only visible in dev mode

> .com site → Upload page → switch to Jetpack site → Back — ❌

For this we can use our dedicated back state. I'll do that in another PR.

> What happens if I browse away (back) while it's uploading?

You can come back and see the progress (try it with a big theme like harmonic). However, the notices only ever show if you happen to be on the page when they happen. I'll look into changing that.

**Comment by seear** [comment]
> Can we make only the success message to disappear, and let the error messages stay?

ok, done.

File: client/my-sites/themes/theme-upload/index.jsx [hunk 1]
```diff
@@ -58,8 +57,7 @@ class Upload extends React.Component {
 		installing: React.PropTypes.bool,
 	};
 
-	constructor( props ) {
-		super( props );
+	componentDidMount() {
```
Comment (omarjackman) [inline client/my-sites/themes/theme-upload/index.jsx]: I don't see any issue here but I'm curious about what prompted this change?
Was the `clearThemeUpload` action not getting called at the right time?

File: client/my-sites/themes/theme-upload/index.jsx [hunk 2]
```diff
@@ -81,12 +79,15 @@ class Upload extends React.Component {
...
-				name: uploadedTheme ? uploadedTheme.name : themeId
-			}
-		} ) );
+		notices.success(
+			translate( 'Successfully uploaded theme %(name)s', {
+				args: {
+					// using themeId lets us show a message before theme data arrives
+					name: uploadedTheme ? uploadedTheme.name : themeId
+				}
+			}, ),
```
Comment (omarjackman) [inline client/my-sites/themes/theme-upload/index.jsx]: This dangling `}, ` looks out of place

File: client/my-sites/themes/theme-upload/index.jsx (hunk 1 above)
Comment (seear) [inline client/my-sites/themes/theme-upload/index.jsx]: There was no issue with `clearThemeUpload` being called at the right time (we want to do it whenever the component is mounted, as long as there is no upload in progress), but it seemed that setting state in the constructor was causing the React warning mentioned in the PR description.

Moving the call to `componentDidMount()` gives us the same behaviour without the warning.

**Pull request merged**